clippy_lints/src/as_slice_instead_of_reference_full_range.rs

```diff
@@ -0,0 +1,70 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::source::snippet_with_applicability;
+use rustc_errors::Applicability;
+use rustc_hir::{Expr, ExprKind, LangItem, Mutability, QPath};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Detects if a full range slice reference is used instead of using the `.as_slice()` method.
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Using the `some_value.as_slice()` method is more explicit then using `&some_value[..]`
+    ///
+    /// ### Example
+    /// ```no_run
+    /// let array: [u8; 4] = [0; 4];
+    /// let slice = &array[..];
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// let array: [u8; 4] = [0; 4];
+    /// let slice = array.as_slice();
+    /// ```
+    #[clippy::version = "1.88.0"]
+    pub AS_SLICE_INSTEAD_OF_REFERENCE_FULL_RANGE,
+    nursery,
+    "Use as slice instead of borrow full range."
+}
+declare_lint_pass!(AsSliceInsteadOfReferenceFullRange => [AS_SLICE_INSTEAD_OF_REFERENCE_FULL_RANGE]);
+
+impl LateLintPass<'_> for AsSliceInsteadOfReferenceFullRange {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &Expr<'_>) {
+        if let ExprKind::AddrOf(_, mutability, borrow) = expr.kind
+            && let ExprKind::Index(name, indexing, _) = borrow.kind
+            && let ExprKind::Struct(qpath, _, _) = indexing.kind
+            && let QPath::LangItem(LangItem::RangeFull, _) = qpath
+        {
+            let mut app = Applicability::MachineApplicable;
+            let snippet = snippet_with_applicability(cx, name.span, "..", &mut app);
+
+            match mutability {
+                Mutability::Not => {
+                    span_lint_and_sugg(
+                        cx,
+                        AS_SLICE_INSTEAD_OF_REFERENCE_FULL_RANGE,
+                        expr.span,
+                        "use `.as_slice()` instead of full range slice",
+                        "try",
+                        format!("{snippet}.as_slice()"),
+                        app,
+                    );
+                },
+                Mutability::Mut => {
+                    span_lint_and_sugg(
+                        cx,
+                        AS_SLICE_INSTEAD_OF_REFERENCE_FULL_RANGE,
+                        expr.span,
+                        "use `.as_mut_slice()` instead of full range slice",
+                        "try",
+                        format!("{snippet}.as_mut_slice()"),
+                        app,
+                    );
+                },
+            }
```

Comment:
```suggestion
            let (msg, sugg) = match mutability {
                Mutability::Not => (
                    "use `.as_slice()` instead of full range slice",
                    format!("{snippet}.as_slice()"),
                ),
                Mutability::Mut => (
                    "use `.as_mut_slice()` instead of full range slice",
                    format!("{snippet}.as_mut_slice()"),
                ),
            };
            span_lint_and_sugg(
                cx,
                AS_SLICE_INSTEAD_OF_REFERENCE_FULL_RANGE,
                expr.span,
                msg,
                "try",
                sugg,
                app,
            );
```